skrub/_join_aggregator.py

```diff
@@ -0,0 +1,354 @@
+        """
+        use_pandas = all(
+            [isinstance(table, pd.DataFrame) for table, _, _ in tables]
+        )
+        use_polars = False
+        if POLARS_SETUP:
+            # we don't mix DataFrame and LazyFrame
+            use_polars = (
+                all(
+                    [isinstance(table, pl.DataFrame) for table, _, _ in tables]
+                )
+                or all(
+                    [isinstance(table, pl.LazyFrame) for table, _, _ in tables]
+                )
+            )
+        if use_pandas:
+            return PandasAssemblingEngine()
+        elif use_polars:
+            return PolarsAssemblingEngine()
+        else:
+            raise NotImplementedError(
+                "Only Pandas or Polars DataFrame are currently supported."
+            )
+
+    @abstractmethod
+    def agg(self, table, cols_to_join, cols_to_agg, ops):
+        pass
+
+    @abstractmethod
+    def join(self, left, right, left_on, right_on):
+        pass
+
+
+def pandas_split_num_categ_cols(table):
+    num_cols = table.select_dtypes("number").columns
+    categ_cols = table.select_dtypes(
+        ["object", "string", "category"]
+    ).columns
+    return num_cols, categ_cols
+    
+
+def pandas_get_agg_ops(cols, agg_ops):
+    pandas_ops_mapping = {
+        "mode": pd.Series.mode
+    }
+    stats = {}
+    for col, op_name in product(cols, agg_ops):
+        op = pandas_ops_mapping.get(op_name, op_name)
+        stats[f"{col}_{op_name}"] = pd.NamedAgg(col, op)
+    return stats
+
+
+class PandasAssemblingEngine(AssemblingEngine):
+
+    def agg(self, table, cols_to_join, cols_to_agg, agg_ops):
+        
+        num_cols, categ_cols = pandas_split_num_categ_cols(
+            table[cols_to_agg]
+        )
+        num_ops, categ_ops = split_num_categ_ops(agg_ops)
+
+        num_stats = pandas_get_agg_ops(num_cols, num_ops)
+        categ_stats = pandas_get_agg_ops(categ_cols, categ_ops)
+                    
+        return table.groupby(cols_to_join).agg(**num_stats, **categ_stats)
+
+    def join(self, left, right, left_on, right_on):
+        
+        return left.merge(
+            right,
+            how="left",
+            left_on=left_on,
+            right_on=right_on,
+        )
+
+
+def polars_split_num_categ_cols(table):
+    
+    num_cols = table.select(
+        pl.col(pl.NUMERIC_DTYPES)
+    ).columns
+    
+    categ_cols = table.select(
+        pl.col(pl.Utf8)
+    ).columns
+    
+    return num_cols, categ_cols
+
+
+def polars_get_agg_ops(cols, agg_ops):
+    stats, mode_cols = [], []
+    for col, op_name in product(cols, agg_ops):
+        op_dict = {
+            "mean": pl.col(col).mean().alias(f"{col}_{op_name}"),
+            "std": pl.col(col).std().alias(f"{col}_{op_name}"),
+            "min": pl.col(col).min().alias(f"{col}_{op_name}"),
+            "max": pl.col(col).max().alias(f"{col}_{op_name}"),
+            "mode": pl.col(col).mode().alias(f"{col}_{op_name}"),
```

Comment:
Are these computed lazily? It seems that we are instantiating many objects that we might not be using. I hope that the instantiation is cheap.

Reply:
We're simply defining operators here, no computation happens, lazy or not. We give these to a table, lazy or not, and let Polars do the heavy lifting for us.